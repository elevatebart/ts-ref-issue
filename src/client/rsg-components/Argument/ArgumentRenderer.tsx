import React from 'react';
import PropTypes from 'prop-types';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import Markdown from 'rsg-components/Markdown';
import Name from 'rsg-components/Name';
import Type from 'rsg-components/Type';
import Group from 'react-group';
import * as Rsg from '../../../typings';

export const styles = ({ space }: Rsg.Theme) => ({
	block: {
		marginBottom: space[2],
	},
});

export interface ArgumentProps {
	name: string;
	type?: any;
	default?: string;
	description?: string;
	returns?: boolean;
	block?: boolean;
}

type ArgumentPropsWithClasses = ArgumentProps & JssInjectedProps;

export const ArgumentRenderer: React.FunctionComponent<ArgumentPropsWithClasses> = ({
	classes,
	name,
	type,
	description,
	returns,
	block,
	...props
}) => {
	const isOptional = type && type.type === 'OptionalType';
	const defaultValue = props.default;
	if (isOptional) {
		type = type.expression;
	}
	const content = (
		<Group>
			{returns && 'Returns'}
			{name && (
				<span>
					<Name>{name}</Name>
					{type && ':'}
				</span>
			)}
			{type && (
				<Type>
					{type.name}
					{isOptional && '?'}
					{!!defaultValue && `=${defaultValue}`}
				</Type>
			)}
			{type && description && `â€”`}
			{description && <Markdown text={`${description}`} inline />}
		</Group>
	);

	if (block) {
		return <div className={classes.block}>{content}</div>;
	}

	return content;
};

ArgumentRenderer.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	name: PropTypes.string.isRequired,
	type: PropTypes.object,
	default: PropTypes.string,
	description: PropTypes.string,
	returns: PropTypes.bool,
	block: PropTypes.bool,
};

export default Styled<ArgumentPropsWithClasses>(styles)(ArgumentRenderer);
