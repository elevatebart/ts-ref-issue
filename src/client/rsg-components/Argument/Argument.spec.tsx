import React from 'react';
import { shallow } from 'enzyme';
import { ArgumentRenderer, styles } from './ArgumentRenderer';

const name = 'Foo';
const type = { name: 'Array' };
const description = 'Converts foo to bar';
const props = {
	classes: classes(styles),
	name: 'argname',
};

it('should render argument', () => {
	const actual = shallow(
		<ArgumentRenderer {...props} name={name} type={type} description={description} />
	);

	expect(actual).toMatchSnapshot();
});

it('should render argument without type', () => {
	const actual = shallow(<ArgumentRenderer {...props} name={name} description={description} />);

	expect(actual).toMatchSnapshot();
});

it('should render optional argument', () => {
	const actual = shallow(
		<ArgumentRenderer
			{...props}
			type={{ type: 'OptionalType', expression: { name: 'Array' } }}
			description={description}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should render default value of argument', () => {
	const actual = shallow(
		<ArgumentRenderer
			{...props}
			type={{ name: 'String' }}
			default="bar"
			description={description}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should render default value of optional argument', () => {
	const actual = shallow(
		<ArgumentRenderer
			{...props}
			type={{ type: 'OptionalType', expression: { name: 'Boolean' } }}
			default="true"
			description={description}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should render argument without description', () => {
	const actual = shallow(<ArgumentRenderer {...props} name={name} type={type} />);

	expect(actual).toMatchSnapshot();
});

it('should render return value', () => {
	const actual = shallow(
		<ArgumentRenderer {...props} type={type} description={description} returns />
	);

	expect(actual).toMatchSnapshot();
});

it('should render with block styles', () => {
	const actual = shallow(<ArgumentRenderer {...props} block />);

	expect(actual).toMatchSnapshot();
});
