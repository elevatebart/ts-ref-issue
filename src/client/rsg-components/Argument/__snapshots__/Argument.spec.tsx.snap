// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render argument 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      Foo
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    Array
  </Styled(Type)>
  —
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render argument without description 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      Foo
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    Array
  </Styled(Type)>
</Group>
`;

exports[`should render argument without type 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      Foo
    </Styled(Name)>
  </span>
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render default value of argument 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      argname
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    String
    =bar
  </Styled(Type)>
  —
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render default value of optional argument 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      argname
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    Boolean
    ?
    =true
  </Styled(Type)>
  —
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render optional argument 1`] = `
<Group
  separator=" "
>
  <span>
    <Styled(Name)>
      argname
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    Array
    ?
  </Styled(Type)>
  —
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render return value 1`] = `
<Group
  separator=" "
>
  Returns
  <span>
    <Styled(Name)>
      argname
    </Styled(Name)>
    :
  </span>
  <Styled(Type)>
    Array
  </Styled(Type)>
  —
  <Markdown
    inline={true}
    text="Converts foo to bar"
  />
</Group>
`;

exports[`should render with block styles 1`] = `
<div
  className="block"
>
  <Group
    separator=" "
  >
    <span>
      <Styled(Name)>
        argname
      </Styled(Name)>
    </span>
  </Group>
</div>
`;
